[mkdocs] docs: add mosaics concept (#846)
NemProject/nem: mkdocs/pages/en/textbook/mosaics.md

# Mosaics

Mosaic
:   A representation of an asset on the NEM blockchain, commonly called tokens on other protocols.
    For example: currencies, licenses, collectibles, access rights, or voting power.

Unlike smart contract-based tokens on other platforms, NEM mosaics are supported directly at the protocol level,
and require no additional coding to use.

Each mosaic defines a new type of asset, and the individual tokens that belong to this type are called _mosaic units_.
Mosaics can represent fungible assets, such as coins, where each unit is interchangeable, and non-fungible assets,
such as paintings or <NFTs:>, where each unit is unique.

A mosaic lives under a registered <namespace:>, is identified by a [fully qualified name](#fully-qualified-name),
and inherits its duration from its parent namespace's lease (see [Lifetime](#lifetime)).

## Name

> The mosaic name is the local identifier for a mosaic within its namespace, and must be unique within it.

Reviewer: ```suggestion
The mosaic name is the identifier for a mosaic within its namespace, and must be unique within it.
```

I think the word "local" is confusing right at the beginning, and it becomes redundant later on when you say "within its namespace".

The name follows specific formatting rules:

> - It can contain lowercase letters, numbers, hyphens `-`, underscores `_`, and apostrophes `'`.

Reviewer: ```suggestion
- It can only contain lowercase letters, numbers, hyphens `-`, underscores `_`, and apostrophes `'`.
```

- It must start with a letter or a number.
- It can be at most **32 characters** long.

Once registered, a mosaic name cannot be changed.

## Fully qualified name

A mosaic's **fully qualified name** (also called its **mosaic ID**) is the unique identifier on the network.
It joins the namespace and the local mosaic name with a colon, in the form `<namespace>:<mosaic name>`:

| Part            | Definition                                          | Length limit                                                |
| --------------- | --------------------------------------------------- | ----------------------------------------------------------- |
| **Namespace**   | One to three dot-separated segments before the `:`. | 16 characters (root) and 64 characters (each subnamespace). |
| **Mosaic name** | The local name after the `:`.                       | 32 characters.                                              |

Namespace segments follow the [namespace name rules](./namespaces.md#name).

Examples:

- `nem:xem`: The native network currency.
- `mycompany.tokens:goldcoin`: A hypothetical company token.

## Description

> Each mosaic has a **description**: a free-text field of up to **512 characters** documenting the asset, for example

Reviewer: ```suggestion
Each mosaic can have a **description**: a free-text field of up to **512 characters** documenting the asset, for example
```

Or is it mandatory? In that case use "must have".

its purpose, origin, or terms of use.

## Properties

Mosaics expose behavioral properties that govern how they can be transferred and how their supply evolves.

### Divisibility

Divisibility
:   Defines how many decimal places a mosaic quantity can have.
    A mosaic with divisibility `0` is indivisible: it can only be transferred in whole units.
    Higher values allow fractional units.

For example, a divisibility of `2` means each _whole unit_ can be divided into 100 _fractional units_ (10^2^),
allowing the mosaic to be handled in increments of `0.01`.

> Fractional units are also called _atomic units_, so in this example, 1 unit consists of 100 atomic units.

Reviewer: ```suggestion
Fractional units are also called _atomic units_, so in this example, 1 whole unit consists of 100 atomic units.
```

Added "whole" just for extra clarity.

In many other protocols, this value is hardcoded.
For example, Bitcoin uses 8 decimal places, and Ethereum uses 18.
NEM allows each mosaic to define its own divisibility, depending on the needs of the asset it represents.

The maximum allowed divisibility in NEM is `6`.

### Initial Supply

Defines the total number of mosaic units created at issuance.

> A mosaic's total supply cannot exceed **9 × 10^15^** atomic units, regardless of divisibility.

Author: We might want to mention the max total supply on Symbol docs.

Reviewer: Agreed!

    rankdir="LR";
    node [fontsize=12];
    "Mosaic Creator";
    "Account A";
    "Account B";

    "Mosaic Creator" -> "Account A" [dir=both];
    "Mosaic Creator" -> "Account B" [dir=both];
    "Account A" -> "Account B" [dir=both style=dashed labeldistance=7 labelangle=-60
    minlen=4 headlabel="Only if mosaic\nis transferable"];

    { rank = same; "Account A"; "Account B"; }
}
```

## Levy

A mosaic definition may include an optional **levy**, a fee paid to a designated account on every transfer of that
mosaic.

A levy specifies four fields:

| Field         | Description                                                                                              |
| ------------- | -------------------------------------------------------------------------------------------------------- |
| **Type**      | `Absolute` (fixed quantity) or `Percentile` (proportional to the amount transferred).                    |
| **Recipient** | Account that receives the levy on every transfer.                                                        |
| **Mosaic ID** | The mosaic in which the levy is paid. It may differ from the mosaic being transferred.                   |

> | **Fee**       | Quantity of the levy mosaic. For `Percentile`, the levy equals `Fee / 10000` of the transferred amount.  |

Reviewer: And if the Type is Absolute, this Fee is in Atomic or Whole units?

Also, worth mentioning that these are called [Basis Points](https://en.wikipedia.org/wiki/Basis_point), or bips.
And I just discovered they have a symbol: ‱ (no need to add the symbol to the docs).

> When a transfer transaction includes a mosaic with a levy, the network charges the levy to the sender and credits it to

Reviewer: We need to clarify:
- Is the levy subtracted from the transferred amount (the recipient receives less), or from the source account (the source pays more than transferred).
- What happens if the source does not have enough funds? I guess it's obvious that the tx fails, but we better be explicit.

the levy recipient, in addition to the regular transaction fee.

For example, a `Percentile` levy with `Fee = 100` charges 1% of the transferred amount.
A transfer of `1 000` atomic units triggers a levy of `10` atomic units credited to the levy recipient.

## Lifetime

A mosaic has no duration of its own.
Its lifetime is tied to the lifetime of its parent namespace:

- While the namespace is active, the mosaic can be transferred and its supply can be changed (subject to its
[properties](#properties)).
- When the namespace expires, the mosaic becomes inactive: transfers and supply changes are rejected, but existing
    balances are preserved.

> - If the original owner renews the namespace during the grace period, the mosaic and its balances become usable again.

Reviewer: Worth linking to what "grace period" means, in the Namespaces page.

> !!! warning "Mosaic loss past the grace period is permanent"

Author: In Symbol, it might be worth investigating and documenting what happens with balances if the account creator tries to renew an expired mosaic.

Reviewer: Done!

    After the grace period, the mosaic is permanently lost.
    Accounts holding the mosaic lose access to their balances and the mosaic cannot be transferred.

    Registering a mosaic with the same name later creates a new asset, not a recovery of the original.

See [Namespace duration](./namespaces.md#duration) for the namespace lease lifecycle.

## Creation Fee

> Creating a mosaic requires paying a one-time creation fee of **10 XEM**.

Reviewer: ```suggestion
Registering a new mosaic requires paying a one-time _creation fee_ of **10 XEM**.
```

To avoid too many "create".

The fee must be paid at the time of creation and is non-refundable.

!!! note "Transaction fee vs. creation fee"
    Creating a mosaic requires announcing a transaction, which also has an associated fee.
    However, this transaction fee is typically negligible compared to the creation fee.

> ## Modifying a Mosaic

Author: We could add a similar section in Symbol docs.

Reviewer: Agreed
